[mideaac] Initial contribution

apella12 opened the pull request:
<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").

ATTENTION: Don't use "git merge" when working with your pull request branch!
This can clutter your Git history and make your PR unusable.
Use "git rebase" instead. See this forum post for further details:
https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358

All PRs should be created using the "main" branch as base.
Important bugfixes are cherry-picked by maintainers to the patch release branch after a PR has been merged.

Add one or more appropriate labels to make your PR show up in the release notes.
E.g. enhancement, bug, documentation, new binding
This can only be done by yourself if you already contributed to this repo.

If your PR's code is not backward compatible with previous releases (which
should be avoided), add a message to the release notes by filing another PR:
https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update.lst

# Title

Provide a short summary in the *Title* above. It will show up in the release notes.
For example:
- [homematic] Improve communication with weak signal devices
- [timemachine][WIP] Initial contribution

# Description

Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work

# Testing

Your pull request will automatically be built and available under the following folder:
https://openhab.jfrog.io/ui/native/libs-pullrequest-local/org/openhab/addons/bundles/

It is a good practice to add a URL to your built JAR in this pull request description,
so it is easier for the community to test your Add-on.
If your pull request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Marketplace:
https://community.openhab.org/c/marketplace/69

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->
This is a second attempt.  The first attempt is closed #17395. Essentially all of the previous comments have been addressed.  The main new change is the separation of the connection manager and the MideaACHandler, (largely rewritten by @lsiepel) these two files are basically new.  I have worked and tested with this version (made some modifications) for several weeks and it seems to work as well as the first attempt now. Also there were a number of new supporting files added by @lsiepel.

For reference the previous intro is here.
I’d like to put this out for new binding review.
This binding was started by the original author as a forum topic but has since had other contributors. I picked it up when I installed a mini-split AC that used the Midea protocol earlier this year, originally to allow for longer polling frequencies. Then as a retirement challenge (no formal java training), decided to attempt to clean it up for official status. I cleaned up the summary report from over 150 items and believe I conformed to current developer guidelines (Java docs, naming, tests, etc.). Along the way added some additional functionality and corrected some code issues.
The discovery and security classes remain largely unchanged from the original author (who I haven’t been able to contact), and work well. The other classes I understand pretty well at this point and hopefully can address any concerns.
As a note, to conform to the guidelines there are breaking changes in naming from the version I published on the forum thread.

apella12 committed:
Midea AC after partial PR review

Mideaac binding after partial PR review.  Main remaining issue is the connection manager which currently needs to be embedded in the MideaACHandler to leverage the OH base thing handler.
Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Working version of split connection manager

Working (sort of) version of split connection manager. Problem with the connection manager being null when the binding is stop/start.  Need reset to clear with clean-cache too.
Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Changes to get separate Connection Manager working

After initial changes, tested various scenarios and made changes so they are working like before.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Apply spotless changes

forgot to run spotless on the last update

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
New PR candidate

Java doc and possible mdns discovery.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Spotless changes

spotless
Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Make retries more robust

Makes retries more robust. Three times for connection and one retry on the command.
Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Align V2 response with V3 reponse

Changed V2 response to align with V3 process after extra decoding.
Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

wborn commented on bundles/org.openhab.binding.mideaac/pom.xml:
```diff
@@ -0,0 +1,17 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>4.3.0-SNAPSHOT</version>
```
```suggestion
    <version>5.0.0-SNAPSHOT</version>
```

apella12 replied:
done

apella12 committed:
Change to OH5.0 snapshot

Changed version and changed dates (in advance-hopefully ok)
Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Changing dates was a bad idea

Changing dates early was a bad idea

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Revert "Changing dates was a bad idea"

This reverts commit 44bb90ff99ff470bfd6efd1ddbcdf3dc8f7f31c9.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

lsiepel commented:
This is not forgotten, the time I have to spare is consumed by the ZUI binding at the moment. So don’t wait for me.

apella12 commented:
@lsiepel Well I'm not really waiting. I'm working on other things too.  There shouldn't be much to review after you rewrote the connection manager. I did a few things to make it more robust to run without the A/C connected to the cloud and fixed some things that you left hanging.  I have had it polling for months without issues.  It is not exactly A/C season, so no one is exactly clamoring for an update.  Hopefully you or someone else can have a look in a couple of months as we get towards spring.

apella12 committed:
Change to new Headers

Change to new Headers
Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

lsiepel commented:
Forgot about this one, sorry. I might have some time to review in the next days.
It was tested and all works as expected?

apella12 commented:
Yes, it is working fine.  I have been using it since the last commit.  Decided to rebase today in case something in the OH5.0 causes a problem, but it passed checks

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,398 @@
+                config.ipPort, config.timeout, config.key, config.token, config.cloud, config.email, config.password,
+                config.deviceId, config.version, config.promptTone);
+
+        startScheduler(2, config.pollingTime, TimeUnit.SECONDS);
```
Reduce complexity and remove the startScheduler method too: 
```suggestion
        scheduledTask = scheduler.scheduleWithFixedDelay(this::pollJob, 2, config.pollingTime, TimeUnit.SECONDS);
```

apella12 replied:
As noted above it has been a while.  However, I hate to mess with this section as I had issues with the original split of the connection manager with multiple schedulers getting started (one reason to kill the thread on line 220).  I need the logic in the startScheduler method.

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,398 @@
+
+                logger.trace("Token: {}", tk.token());
+                logger.trace("Key: {}", tk.key());
+                logger.info("Token and Key obtained from cloud, saving, back to initialize");
```
Prevend logging to info at all places
```suggestion
                logger.debug("Token and Key obtained from cloud, saving, back to initialize");
```

apella12 replied:
ok

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,398 @@
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, String.format(
+                        "Can't retrieve Token and Key from Cloud; email, password and/or cloud parameter error"));
+                logger.warn("Can't retrieve Token and Key from Cloud; email, password and/or cloud parameter error");
```
this logging is not needed when the thing state is set.
```suggestion
```

apella12 replied:
See question above

apella12 replied:
Ok in next commit

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/connection/ConnectionManager.java:
```diff
@@ -0,0 +1,540 @@
+     * @param str string to be evaluated
+     * @return boolean true or false
+     */
+    public static boolean isBlank(String str) {
+        return str.trim().isEmpty();
+    }
```
Remove this and use `String.isBlank()` wherever it is called.

apella12 replied:
ok

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/connection/ConnectionManager.java:
```diff
@@ -0,0 +1,540 @@
+        }
+        if (retryCount == maxTries) {
+            deviceIsConnected = false;
+            logger.info("Failed to connect after {} tries. Try again with next scheduled poll", maxTries);
```
```suggestion
            logger.debug("Failed to connect after {} tries. Try again with next scheduled poll", maxTries);
```

apella12 replied:
Ok.  Had that for testing, forgot to take out

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/connection/ConnectionManager.java:
```diff
@@ -0,0 +1,540 @@
+                    try {
+                        Thread.sleep(5000);
+                    } catch (InterruptedException ex) {
+                        logger.debug("An interupted error (socket retry) has occured {}", ex.getMessage());
```
Stop processing and exit immediate when an InterruptException occurs.

apella12 replied:
As I recall that won't work as I intended (with 3 tries).  The exceptions aren't serious and ephemeral in nature.  I haven't had that INFO log message in months (after I went from 2 to three tries.  That's why I forgot it was there.

apella12 replied:
I have separated the catches

lsiepel reviewed (changes requested):
Thanks for not leaving this PR/binding. It has improved, i also see more room for improvement. Besides some new minor issues that i probably missed before, i think the DTO and connection structure can benefit from some more work and thinking.

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/MideaACConfiguration.java:
```diff
@@ -0,0 +1,124 @@
+     * @return true(discovery needed), false (not needed)
+     */
+    public boolean isDiscoveryNeeded() {
+        return ("0".equalsIgnoreCase(deviceId) || deviceId.isBlank() || ipPort <= 0 || ipAddress.isBlank()
```
OK

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/MideaACConfiguration.java:
```diff
@@ -0,0 +1,124 @@
+     * @return true (yes they can), false (they cannot)
+     */
+    public boolean isTokenKeyObtainable() {
+        return (!email.isBlank() && !password.isBlank() && !"".equals(cloud));
```
Ok

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/MideaACConfiguration.java:
```diff
@@ -0,0 +1,124 @@
+     * @return true (Valid, all items are present) false (key, token and/or provider missing)
+     */
+    public boolean isV3ConfigValid() {
+        return (!key.isBlank() && !token.isBlank() && !"".equals(cloud));
```
Ok

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/Utils.java:
```diff
@@ -0,0 +1,250 @@
+    public static String bytesToBinary(byte[] bytes) {
+        String s1 = "";
+        for (int j = 0; j < bytes.length; j++) {
+
```
ok

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/Utils.java:
```diff
@@ -0,0 +1,250 @@
+
+            s1 = s1.concat(Integer.toBinaryString(bytes[j] & 255 | 256).substring(1));
+            s1 = s1.concat(" ");
+
```
OK

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,258 @@
+	<channel-type id="power">
+		<item-type>Switch</item-type>
+		<label>Power</label>
+		<description>Turn the AC on and off.</description>
```
OK

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/Utils.java:
```diff
@@ -0,0 +1,250 @@
+     * @param bytes bytes to convert
+     * @return string of hex chars
+     */
+    public static String bytesToHex(byte[] bytes) {
```
I inherited this from the Initial Contribution.  Will need study

apella12 replied:
Ok figured this out. Testing now, but should be in next commit

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/Utils.java:
```diff
@@ -0,0 +1,250 @@
+        return s1;
+    }
+
+    /**
+     * Converts byte array to lower case hex string
+     * 
+     * @param bytes bytes to convert
+     * @return string of hex chars
+     */
+    public static String bytesToHexLowercase(byte[] bytes) {
+        char[] hexChars = new char[bytes.length * 2];
+        for (int j = 0; j < bytes.length; j++) {
+            int v = bytes[j] & 0xFF;
+            hexChars[j * 2] = HEX_ARRAY_LOWERCASE[v >>> 4];
+            hexChars[j * 2 + 1] = HEX_ARRAY_LOWERCASE[v & 0x0F];
+        }
+        return new String(hexChars);
+    }
```
I inherited this from the Initial Contribution. Will need study

apella12 replied:
Ok figured this out. Testing now, but should be in next commit

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/Utils.java:
```diff
@@ -0,0 +1,250 @@
+     * @param s string to convert to byte array
+     * @return byte array
+     */
+    public static byte[] hexStringToByteArray(String s) {
+        int len = s.length();
+        byte[] data = new byte[len / 2];
+        for (int i = 0; i < len; i += 2) {
+            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4) + Character.digit(s.charAt(i + 1), 16));
+        }
+        return data;
+    }
```
I inherited this from the Initial Contribution. Will need study

apella12 replied:
Ok figured this out. Testing now, but should be in next commit

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,258 @@
+	<channel-type id="target-temperature">
+		<item-type>Number:Temperature</item-type>
+		<label>Target temperature</label>
+		<description>Target temperature.</description>
```
OK

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,258 @@
+	</channel-type>
+	<channel-type id="target-temperature">
+		<item-type>Number:Temperature</item-type>
+		<label>Target temperature</label>
```
Think I have them checked in next draft

apella12 commented on bundles/org.openhab.binding.mideaac/README.md:
```diff
@@ -0,0 +1,120 @@
+
+## Discovery
+
+Once the Air Conditioner is on the network (WiFi active) the other required parameters can be discovered automatically.
```
Slight change; Once the Air Conditioner is on the network (WiFi active) most required parameters will be discovered automatically.  The nuance is that deviceId, token, key and version are discovered automatically if you added the cloud and password (unless V2), you still need to enter pollingTime and socket timeout and prompt tone (but they do have fault values).  Not really sure what you want here.

lsiepel replied:
Here you provide a lot more information then in the readme. This is what chatgpt makes out of it. Please check and adapt the text so that it gives the user enough guidance to set it up correctly.l

```
The key parameters—like deviceId, token, key, and version—are automatically discovered once the air conditioner is connected to the WiFi and you’ve provided the necessary cloud and password (unless it's a version 2 device).

However, you’ll still need to manually enter a few parameters such as: Polling time, Socket timeout, Prompt tone.
```

apella12 replied:
Made more of a step-by-step process in next commit.

apella12 commented on bundles/org.openhab.binding.mideaac/README.md:
````diff
@@ -0,0 +1,120 @@
+```java
+sitemap midea label="Split AC MBR"{
+    Frame label="AC Unit" {
+    Text item=outdoor_temperature label="Outdoor Temperature [%.1f °F]"
+    Text item=indoor_temperature label="Indoor Temperature [%.1f °F]"
+    Setpoint item=target_temperature label="Target Temperature [%.1f °F]" minValue=63.0 maxValue=78 step=1.0
+    Switch item=power label="Midea AC Power"
+    Switch item=temperature_unit label= "Temp Unit" mappings=[ON="Fahrenheit", OFF="Celsius"]
+    Selection item=fan_speed label="Midea AC Fan Speed"
+    Selection item=operational_mode label="Midea AC Mode"
+    Selection item=swing_mode label="Midea AC Louver Swing Mode"
+}
````
I moved the curly bracket out, if that was what you meant

lsiepel replied:
Created a new comment with a suggestion. I meant to add indentation not removing a curly brace.

apella12 replied:
I didn't take it out, just tabbed it over.  Will look at new comment (I'm working from the top down)

apella12 replied:
ok.  Should be in last commit

apella12 commented on bundles/org.openhab.binding.mideaac/README.md:
````diff
@@ -0,0 +1,120 @@
+}
+}
+```
+
+## Debugging and Tracing
+
+Switch the log level to TRACE or DEBUG on the UI Settings Page (Add-on Settings)
````
ok

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/security/Security.java:
```diff
@@ -0,0 +1,627 @@
+            md.update(combine);
+            return md.digest();
+        } catch (NoSuchAlgorithmException e) {
+        }
```
ok

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/dto/CloudDTO.java:
```diff
@@ -0,0 +1,357 @@
+        String json = data.toString();
+        logger.debug("Request json: {}", json);
+
+        Request request = httpClient.newRequest(url).method(HttpMethod.POST).timeout(15, TimeUnit.SECONDS);
```
This was from the initial contributor.  There are several back and forth JSONs. I don't think there is a lot of logic except to satisfy the cloud connection.  I'm not going to be able to change this.  If it is a deal stopper, I'll just leave as a forum binding, sorry.

apella12 replied:
Eliminated the dto folder and DTO classes.  Made Cloud nonNull.

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/discovery/MideaACDiscoveryService.java:
```diff
@@ -0,0 +1,354 @@
+            final DatagramPacket receivePacket = new DatagramPacket(buffer, buffer.length);
+            // No need to call close first, because the caller of this method already has done it.
+            startDiscoverSocket();
+            // Runs until the socket call gets a time out and throws an exception. When a time out is triggered it means
+            // no data was present and nothing new to discover.
+            while (true) {
```
Probably, but also inherited code

lsiepel replied:
Will proceed for now, this can be adapted later on.

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/discovery/Connection.java:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * UDP port1 to send command.
+     */
+    public static final int MIDEAAC_SEND_PORT1 = 6445;
+    /**
+     * UDP port2 to send command.
+     */
+    public static final int MIDEAAC_SEND_PORT2 = 20086;
+    /**
+     * UDP port devices send discover replies back.
+     */
+    public static final int MIDEAAC_RECEIVE_PORT = 6440;
```
Don't really know.

lsiepel replied:
If the discovery works as expected we leave this as is

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,398 @@
+        config = getConfigAs(MideaACConfiguration.class);
+
+        if (!config.isValid()) {
+            logger.warn("Configuration invalid for {}", thing.getUID());
```
I don't understand this comment.  Are you saying that because the Thing is INITIALIZING we don't need logging?

lsiepel replied:
The logging is duplicate information when the Thing state is also set. So remove the logging.

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,398 @@
+        if (!config.isValid()) {
+            logger.warn("Configuration invalid for {}", thing.getUID());
+            if (config.isDiscoveryNeeded()) {
+                logger.warn("Discovery needed, discovering....{}", thing.getUID());
```
Same question as above

lsiepel replied:
```suggestion
```

apella12 replied:
Ok in next commit

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,398 @@
+                MideaACDiscoveryService discoveryService = new MideaACDiscoveryService();
+
+                try {
+                    discoveryService.discoverThing(config.ipAddress, this);
```
IT has been months since I have studied this code.  Will need to think about this.  It has always been there.

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,398 @@
+                    return;
+                }
+            } else {
+                logger.debug("MideaACHandler config of {} is invalid. Check configuration", thing.getUID());
```
See above question?

apella12 replied:
Ok in next commit

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,398 @@
+                logger.info("Retrieving Token and/or Key from cloud");
+                CloudProviderDTO cloudProvider = CloudProviderDTO.getCloudProvider(config.cloud);
+                getTokenKeyCloud(cloudProvider);
+                return;
```
I put that in to stop that thread since the get tokenKeyCloud calls initialize().

lsiepel replied:
Not a preferred pattern. The tolenKeyoud should do the token stuff and not start a initialize. That method should only be called by the framework

apella12 replied:
I did have one thought on this.  If I make any change to the UI page, including the location (unrelated to the device itself), initialize() is triggered.  I don't understand why it is not automatically triggered when the key and token are updated from the internet and saved. (hence the need for the call)? Also without the call the new token and key aren't passed back and updated (authorization fails even though the UI shows token and key).

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,398 @@
+                getTokenKeyCloud(cloudProvider);
+                return;
+            } else {
+                logger.warn("Configuration invalid for {} and no account info to retrieve from cloud", thing.getUID());
+                return;
```
Ok I think

apella12 committed:
Address Comments partial

Addressed some PR comments. Some questions outstanding.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,399 @@
+                connectionManager.sendCommand(CommandHelper.handleOffTimer(command, lastresponse), callbackLambda);
+            }
+        } catch (MideaConnectionException | MideaAuthenticationException e) {
+            logger.warn("Unable to proces command: {}", e.getMessage());
```
When an authentication exception occurs, i guess it is permanenet. When a timeout occurs, it is transient and the binding should recover. You might want to handle both cases seperate.
```suggestion
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
```

apella12 replied:
Ok This was part of the code you rewrote and have never seen an exception, so don't know what else to say.

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/connection/ConnectionManager.java:
```diff
@@ -0,0 +1,530 @@
+            writer.close();
+            inputStream.close();
+            socket.close();
+
```
```suggestion
```

apella12 replied:
I think I have all these addressed in next commit

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/connection/ConnectionManager.java:
```diff
@@ -0,0 +1,530 @@
+                } else if (Arrays.equals(new String("ERROR").getBytes(), response)) {
+                    throw new MideaAuthenticationException("Authentication failed!");
+                } else {
+                    logger.warn("Authentication reponse unexpected data length ({} instead of 72)!", response.length);
```
As the exception bubbles up and should set the thing state, it would be enough to log this into debug.
```suggestion
                    logger.debug("Authentication reponse unexpected data length ({} instead of 72)!", response.length);
```

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/connection/ConnectionManager.java:
```diff
@@ -0,0 +1,530 @@
+                                    callback.updateChannels(lastResponse);
+                                }
+                            } catch (Exception ex) {
+                                logger.warn("Processing response exception: {}", ex.getMessage());
```
What kind of exception do we expect here? This catches all kind of exceptions. I guess we should throw a `MideaException` here?

apella12 replied:
I got an error message when with that 
![image](https://github.com/user-attachments/assets/99bd79ec-f0ef-4752-a9bc-57c90b255df7)

lsiepel replied:
Throw not catch

apella12 replied:
Think I have that fixed

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,398 @@
+                    discoveryService.discoverThing(config.ipAddress, this);
+                    return;
+                } catch (Exception e) {
+                    logger.error("Discovery failure for {}: {}", thing.getUID(), e.getMessage());
```
Change it to debug as the message might contain valueable information for debugging.
```suggestion
                    logger.debug("Discovery failure for {}: {}", thing.getUID(), e.getMessage());
```

apella12 replied:
ok

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/MideaACConfiguration.java:
```diff
@@ -0,0 +1,124 @@
+     * @return true (Valid, all items are present) false (key, token and/or provider missing)
+     */
+    public boolean isV3ConfigValid() {
+        return (!key.isBlank() && !token.isBlank() && !"".equals(cloud) && !cloud.isBlank());
```
The `isBlank()` covers this. No need to double up.
```suggestion
        return (!key.isBlank() && !token.isBlank() && !cloud.isBlank());
```

apella12 replied:
ok

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/MideaACConfiguration.java:
```diff
@@ -0,0 +1,124 @@
+     * @return true (yes they can), false (they cannot)
+     */
+    public boolean isTokenKeyObtainable() {
+        return (!email.isBlank() && !password.isBlank() && !"".equals(cloud) && !cloud.isBlank());
```
```suggestion
        return (!email.isBlank() && !password.isBlank() && !cloud.isBlank());
```

lsiepel commented on bundles/org.openhab.binding.mideaac/README.md:
```diff
@@ -0,0 +1,116 @@
+
+| Channel                      | Type               | Description                                                                                            | Read only | Advanced |
+|--:---------------------------|--:-----------------|--:-----------------------------------------------------------------------------------------------------|--:--------|--:-------|
+| power                        | Switch             | Turn the AC on and off.                                                                                |           |          |
```
```suggestion
| power                        | Switch             | Turn the AC on or off.                                                                                 |           |          |
```

apella12 replied:
ok

lsiepel commented on bundles/org.openhab.binding.mideaac/README.md:
````diff
@@ -0,0 +1,116 @@
+
+```java
+sitemap midea label="Split AC MBR"{
+    Frame label="AC Unit" {
+    Text item=outdoor_temperature label="Outdoor Temperature [%.1f °F]"
+    Text item=indoor_temperature label="Indoor Temperature [%.1f °F]"
+    Setpoint item=target_temperature label="Target Temperature [%.1f °F]" minValue=63.0 maxValue=78 step=1.0
+    Switch item=power label="Midea AC Power"
+    Switch item=temperature_unit label= "Temp Unit" mappings=[ON="Fahrenheit", OFF="Celsius"]
+    Selection item=fan_speed label="Midea AC Fan Speed"
+    Selection item=operational_mode label="Midea AC Mode"
+    Selection item=swing_mode label="Midea AC Louver Swing Mode"
````
```suggestion
    Frame label="AC Unit" {
        Text item=outdoor_temperature label="Outdoor Temperature [%.1f °F]"
        Text item=indoor_temperature label="Indoor Temperature [%.1f °F]"
        Setpoint item=target_temperature label="Target Temperature [%.1f °F]" minValue=63.0 maxValue=78 step=1.0
        Switch item=power label="Midea AC Power"
        Switch item=temperature_unit label= "Temp Unit" mappings=[ON="Fahrenheit", OFF="Celsius"]
        Selection item=fan_speed label="Midea AC Fan Speed"
        Selection item=operational_mode label="Midea AC Mode"
        Selection item=swing_mode label="Midea AC Louver Swing Mode"
```

apella12 replied:
ok

lsiepel reviewed (changes requested):
Marked some previous comments as resolved. Found some new ones, but nothing big. Let's slowly continue and narrow the remaining issues down and fix them together.

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/connection/ConnectionManager.java:
```diff
@@ -0,0 +1,530 @@
+            }
+        }
+
+        if (!deviceIsConnected) {
+            logger.info("Connected to IP {}", ipAddress);
+        }
+        logger.debug("Connected to IP {}", ipAddress);
```
What the intention here was one INFO for the first time and debug if it was recovering from some problem, but not disconnected. Otherwise there is nothing in the log to show it is running.  Note starting date.
`2025-01-24 16:45:59.796 [INFO ] [nternal.connection.ConnectionManager] - Connected to IP 192.168.0.246`
This is when I updated to the rebased binding
`2025-02-19 10:58:07.986 [INFO ] [nternal.connection.ConnectionManager] - Connected to IP 192.168.0.246`
This is when I deleted the email and password this morning to recheck that it could be done
`2025-02-20 08:54:25.405 [INFO ] [nternal.connection.ConnectionManager] - Connected to IP 192.168.0.246`

apella12 replied:
Added a comment to explain

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/connection/ConnectionManager.java:
```diff
@@ -0,0 +1,530 @@
+                    sendCommand(command, callback);
+                } else {
+                    droppedCommands = droppedCommands + 1;
+                    logger.info("Problem with reading response, skipping {} skipped count since startup {}", command,
```
ok.  I used that to track health, but probably no value to user

apella12 committed:
Additional edits

Additional edits.
Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 commented:
In no particular order, 1) I'm not sure how to use the OH standard conversions and where to put the .toLowerCase since I do not see that in the documentation. 2) The exception in the retry I'm trying to address is simply a timeout. 3) Elsewhere, I was trying to avoid breaking something that I fixed earlier would cause me to go back and fix again.  It has been months and I'm not recalling all the issues I had getting to the fix. 4) I also attached a log of the DTO actions, it appears to me everything is needed.  The only logic is to pull from the cloud responses to compose the next message.

apella12 committed:
Update utilities and connection manager

Updated utilities to leverage OH core and separated retry timeout from other exceptions.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 commented:
Figured out how to leverage the OH core util hex/byte and separated the socket timeout exception from anything else.  Seems to still work, but with the magnitude of changes am testing overnight. Have tested scan, retrieve key/token, normal polling and temperature changes. Had one timeout exception that worked as expected. 9 seconds = 4 for timeout and 5 delay for next try.
```
2025-02-21 17:19:56.320 [DEBUG] [nternal.connection.ConnectionManager] - Disconnecting from device at 192.168.0.246
2025-02-21 17:20:05.326 [DEBUG] [nternal.connection.ConnectionManager] - Socket retry count 1, Socket timeout connecting to 192.168.0.246: Connect timed out
2025-02-21 17:20:05.329 [DEBUG] [nternal.connection.ConnectionManager] - Device at IP: 192.168.0.246 requires authentication, going to authenticate
```
Normal connection is less that 10ms.
```
2025-02-21 17:09:53.751 [DEBUG] [nternal.connection.ConnectionManager] - Disconnecting from device at 192.168.0.246
2025-02-21 17:09:53.757 [DEBUG] [nternal.connection.ConnectionManager] - Device at IP: 192.168.0.246 requires authentication, going to authenticate
```

apella12 committed:
Eliminate DTO folder and classes

Eliminated DTO folder and classes by addressing null issues in the cloud communications. Plus slight timing issue on retry.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Update pom with spotless

New standard format?

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Change connection log

Change connection log to either/or avoiding both on device being connected.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Minor tweaks from extended testing

Minor tweaks from extended testing

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Add scheduled token and key updates

Add capability for scheduled token and key updates.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Validated the automatic token key update steps

Completed the automatic token key addition with clean-up, additional testing, and eliminated the Clouds class.  Developed additional understanding of the token key update to improve the comments. Corrected an order issue in Cloud Provider.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Cleanup unused code in Cloud.java

Cleanup unused code in Cloud.java

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Add the capability command

These changes add the get capability command and sends it to the device during initial initiation, reads the results and populates the properties for easy reference.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Reorganized files

Reorganized files and general clean-up of capabilities.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Add default cloud

Use the default cloud for faster AC discovery.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
General Cleanup and documentation

Conformed capabilities (properties) to OH standards, added documentation. Changed the ReadMe to fully reflect default options. Changed the text configuration to require IP address.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Document clarifications

General review for document clarifications

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
spotless missed

Forgot to run spotless first
Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Add capability follow-up and energy Poll

Added capability follow-up command and energy polling, command and reporting. Streamlined response() coding between version 2 and 3. Eliminated alternate target temp for now.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Add energy Polling Refresh

Add energy Polling Refresh, convert frequency to minutes. Edits for clarity

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Improve config descriptions

Improve config descriptions and inline documentation

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Correct LED command and change energy polling

Correct LED command (misread python byte) and change energy polling, even when device is off.  There is slight wattage when off to capture.
Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Add target humidity support for Midea AC binding

Introduces a new 'target humidity' channel and associated logic for setting and handling target humidity in dry mode. Adds handling for unsolicited humidity reports (0xA0), updates the command and response processing, extends the callback interface, and provides tests for the new functionality. Updates i18n and thing-type definitions to reflect the new channel.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Add support for maximum humidity and filter status channels

Replaces 'target-humidity' with 'maximum-humidity' channel and adds 'filter-status' channel for Midea AC binding. Implements handling for new unsolicited temperature (0xA1) and humidity (0xA0, 0xC1) responses, updates command and response processing, and extends tests for new features. Updates documentation and i18n resources to reflect these changes.
Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 commented:
@lsiepel I didn't see your reopening comment.  Must not have been notified (or thought it was related to my ZUI nofications)...

To recap; you wrote code to separate the connection manager and the MideaAC handler and I got it working again and addressed all the comments from February 21 and before.   Since then I extracted several features from the HA version (and with the help of Copilot in VSC converted them to java). Also there have been a couple of new users of the forum version that had new messages that incorporated using some Lua code references.  also because of the new features reworked some part of the code.  All that is complete now, but the binding is 25% longer. I was just making a final pass to ensure the ReadMe matched with the additional capabilities

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/security/Crc8.java:
```diff
@@ -0,0 +1,78 @@
+    public static int calculate(byte[] bytes) {
+        int crcValue = 0;
+        for (byte m : bytes) {
+            int k = (byte) (crcValue ^ m);
+            if (k > 256) {
+                k -= 256;
+            }
+            if (k < 0) {
+                k += 256;
+            }
```
Changed this in latest commit

apella12 committed:
Refactor and update Midea AC binding documentation and code

Improved documentation in README and JavaDoc comments, clarified terminology (e.g., 'Maximum Humidity'), and updated test method naming for consistency. Fixed CRC8 calculation logic for correctness. Updated thing-types.xml to use correct state patterns for energy channels. Removed outdated example from README and added clarifying notes about channel support.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Minor clean-ups

Minor cleanups of Temperature response and energy formats.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Update palm to 5.1

Update pom to 5.1

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Add tags to channels

Saw that there was a project to add tags to channels, so tried to follow examples from other bindings.  Don't have sematic model deployed, so wanted to add to the review process.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Add CODEOWNER for mideaac binding

Assigned @apella12 as the code owner for /bundles/org.openhab.binding.mideaac/ in the CODEOWNERS file.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/MideaACConfiguration.java:
```diff
@@ -0,0 +1,146 @@
+     * @return true(discovery needed), false (not needed)
+     */
+    public boolean isDiscoveryPossible() {
+        return (Utils.validateIP(ipAddress));
```
```suggestion
        return Utils.validateIP(ipAddress);
```

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/Utils.java:
```diff
@@ -0,0 +1,253 @@
+     * @param bytes bytes to convert
+     * @return string of upper case hex chars
+     */
+    public static String bytesToHex(byte[] bytes) {
```
The content of this method was adapted according to https://github.com/openhab/openhab-addons/pull/17749/files#r1961996508

Now it is just a proxy and everywhere it is called it should be replaced by the core variant directly.

apella12 replied:
I don't get the reference. It clicks back my PR 17749

lsiepel replied:
The link is referring to thisPR and it shows the file it is about, you might need to click 'load file' if it is to large to show the comment.
Anyway, it looks like all these occurences have been adapted in the mean time as i sugested. thanks

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/cloud/Cloud.java:
```diff
@@ -0,0 +1,334 @@
+            String msg = result.get("msg").getAsString();
+            if (code != 0) {
+                errMsg = msg;
+                logger.warn("Error {} logging to Cloud: {}", code, msg);
```
This logger shows that the interaction with this class and its callers can be improved.
How would the caller know that the login failed and route that to the thing lifecycle? 
 
Usually these type of classes throw a custom exception like `LoginFailedException` or `TimeOutException`. 
Either way it is good to make a distinction between transient and permanent problems so the caller can act on it.

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/connection/ConnectionManager.java:
```diff
@@ -0,0 +1,597 @@
+        // Suspect that the AC device needs a few seconds to clear.
+        if (!resend) {
+            try {
+                Thread.sleep(5000);
```
Noticed two things:
1. the condition seems the reverse. currently it sleeps when not resending
2. a hardcoded sleep is a code smel, what other options did you consider?
3. both a synchronized method and a long (5 second) sleep are error prone.

lsiepel replied:
This one was left open.

lsiepel replied:
My point was that i expect the sleep when (resend == true), but the current condition is the opposite (resend == false)

apella12 replied:
Reviewed. If the resend is true (as it normally is) and the response bytes are still null it hits this `else` (line 401) to start the retry.
```
            } else {
                if (resend) {
                    logger.debug("Resending Command {}", command);
                    resend = false;
                    sendCommand(command, callback);
```
The resend is set to false, so I needed a delay before it tries to create a new socket. As noted elsewhere, I had trouble with an immediate resend. The resend doesn't turn to true again until a successful poll (line 359)
```
            if (responseBytes != null) {
                resend = true;
```

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/connection/ConnectionManager.java:
```diff
@@ -0,0 +1,597 @@
+                if (retrySocket < maxTries) {
+                    try {
+                        Thread.sleep(5000);
+                    } catch (InterruptedException ex) {
```
Not fully fixed this comment: https://github.com/openhab/openhab-addons/pull/17749/files#r1962228100
The interupt flag needs to be restored and exit quickly

apella12 replied:
Made a change for review

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/connection/ConnectionManager.java:
```diff
@@ -0,0 +1,597 @@
+            } catch (InterruptedException e) {
+                logger.debug("An interupted error (write command2) has occured {}", e.getMessage());
+                Thread.currentThread().interrupt();
+                // Note, but continue anyway for second write if needed.
```
when interrupted it should stop processing and not retrying.

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,394 @@
+			<channel id="humidity" typeId="humidity"/>
+			<channel id="kilowatt-hours" typeId="kilowatt-hours"/>
+			<channel id="amperes" typeId="amperes"/>
+			<channel id="watts" typeId="watts"/>
```
Additional to https://github.com/openhab/openhab-addons/pull/17749#discussion_r2479147336

apella12 replied:
Made changes for review

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,394 @@
+				<description>Secret Token (length 128 HEX) used for secure connection authentication used with devices v3 (if not
+					known, enter email, password for Cloud account to retrieve it).</description>
+			</parameter>
+			<parameter name="key" type="text" required="false">
```
I think we can drop this as it is retrieved by the binding, there is no need to make this a configuration option.

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,394 @@
+				<description>Poll device status. Minimum is 30, default 60.</description>
+				<default>60</default>
+			</parameter>
+			<parameter name="energyPoll" type="decimal" required="true" min="0" unit="m">
```
Can be marked advanced. (also applies to some other configuration options)

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,394 @@
+				<description>Version 3 requires Token, Key and Cloud provider. Version 2 doesn't.</description>
+				<default>0</default>
+			</parameter>
+			<parameter name="energyDecode" type="boolean" required="true">
```
Would it be possible to detect this while decoding? It would make this parameter obsolete.

apella12 replied:
Not sure which one you are asking about. The version (which is decoded in discovery) or the energyDecode which is not. HA has a flag I had to try both methods to figure out which one.

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,613 @@
+                MideaACDiscoveryService discoveryService = new MideaACDiscoveryService();
+
+                try {
+                    discoveryService.discoverThing(config.ipAddress, this);
```
This previous comment has not been addressed:
https://github.com/openhab/openhab-addons/pull/17749/files#r1962212773

apella12 replied:
I'm having problems with links to previous comments. I recall adding a catch in this area to address, Not enough?

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,613 @@
+            updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.CONFIGURATION_PENDING, "Configuration not valid");
+            if (config.isDiscoveryPossible()) {
+                updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.CONFIGURATION_PENDING,
+                        "Configuration missing, discovery needed. Discovering...");
```
Please use i18n custom keys when setting the thing state. This would make it possible to add translations.
For reference: https://next.openhab.org/docs/developer/utils/i18n.html#using-custom-keys

(applies to all occurences of `updateStatus` with a thingstatusdetail)

apella12 replied:
Just to check. I have run `mvn i18n:generate-default-translations -pl :org.openhab.binding.mideaac`, but need to manually add the thing status details in the MideaACHandler.

apella12 replied:
I think I have this sorted out. With the limited slate of states and descriptions did some consolidation and removed some unnecessary ones. I have assumed that `e.getMessage()` override the content in the i18n.

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,613 @@
+            if (config.isTokenKeyObtainable()) {
+                try {
+                    CloudProvider cloudProvider = CloudProvider.getCloudProvider(config.cloud);
+                    getTokenKeyCloud(cloudProvider);
```
Blocking call should be run async

lsiepel reviewed (changes requested):
Sorry i had you waiting this long. I got distracted many times and needed some focus time to finish this review. 
Finished all files. The handler might need some refactoring to create logical units that do a strict amount of work. especially around the initialize method and how it is called from other places.

Anyway, i think that depending on the follow-up there might be one more review, but as i have covered everything, it will not take as long as before.

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/MideaACConfiguration.java:
```diff
@@ -0,0 +1,146 @@
+    /**
+     * Cloud Account Password
+     */
+    public String password = "password1";
```
Don't understand. This is used to get the token and key
```
    public boolean isTokenKeyObtainable() {
        return (!email.isBlank() && !password.isBlank() && !cloud.isBlank());
    }
```

lsiepel replied:
Are you saying that all devices have `Password1` as default password? I'm not asking you to remove the field `password`, but suggesting to remove the default value for that field.

```suggestion
    public String password = "";
```

apella12 replied:
Yes "password1" with "nethome+us@mailinator.com" will retrieve the token and key for any device.

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/MideaACConfiguration.java:
```diff
@@ -0,0 +1,146 @@
+    /**
+     * Cloud Account email
+     */
+    public String email = "nethome+us@mailinator.com";
```
Don't understand. This is used to get the token and key (see above) and in handler 252 to 263
        // Check for valid token and key and/or contact cloud account to get them
```
        if (config.version == 3 && !config.isV3ConfigValid()) {
            if (config.isTokenKeyObtainable()) {
                try {
                    CloudProvider cloudProvider = CloudProvider.getCloudProvider(config.cloud);
                    getTokenKeyCloud(cloudProvider);
                    return;
                } catch (Exception e) {
                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
                            "Token and key could not be obtained from Cloud");
                    return;
                }
```

apella12 replied:
Yes "password1" with "nethome+us@mailinator.com" will retrieve the token and key for any device.

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/connection/ConnectionManager.java:
```diff
@@ -0,0 +1,597 @@
+        ensureConnected();
+
+        if (command instanceof CommandSet) {
+            ((CommandSet) command).setPromptTone(promptTone);
```
Done

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/CommandBase.java:
```diff
@@ -0,0 +1,314 @@
+            sum = (byte) (sum + value);
+        }
+        sum = (byte) ((255 - (sum % 256)) + 1);
+        return (byte) sum;
```
done

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,394 @@
+				<description>Password for cloud account chosen in Cloud Provider.</description>
+				<default>password1</default>
+			</parameter>
+			<parameter name="token" type="text" required="false">
```
There are other ways outside of OH to get token and key (there is a HA python script) Also in the case the cloud goes down in the future, a previous token and key (stored elsewhere) can be added to restore operations. They are edge cases, but are easier to leave in now than have to add later.

lsiepel replied:
Do you know what the lifetime is of those tokens? roughly?

apella12 replied:
AFAIK they never expire. I have used for months and reused them.

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,394 @@
+				<description>Secret Key (length 64 HEX) used for secure connection authentication used with devices v3 (if not
+					known, enter email, password for Cloud account to retrieve it).</description>
+			</parameter>
+			<parameter name="pollingTime" type="decimal" required="true" min="30" unit="s">
```
Good point. I don't have an example is this correct?
`<parameter name="token" type="text" required="false" advanced="true">`

lsiepel replied:
No, a child tag
For reference: https://www.openhab.org/docs/developer/addons/config-xml.html#xml-structure-for-configuration-descriptions

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,394 @@
+					known, enter email, password for Cloud account to retrieve it).</description>
+			</parameter>
+			<parameter name="pollingTime" type="decimal" required="true" min="30" unit="s">
+				<context>pollingTime</context>
```
Don't understand. Polling is the main thing going on.

apella12 replied:
One question, are the context needed. Not sure where they came from. Will give a few a shot

lsiepel replied:
it is optional, they are helpfull for the browser / ui / controls.

apella12 replied:
After reviewing the documents most can be removed. Left the unambiguous ones. Done in next commit

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,394 @@
+				<description>Energy polling frequency. default 0 ; (in case not supported).</description>
+				<default>0</default>
+			</parameter>
+			<parameter name="keyTokenUpdate" type="decimal" required="false" min="0" unit="d">
```
Note comment above. Token/key are permanent and reusable

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,613 @@
+                CommandSet humidityUpdate = new CommandSet();
+                humidityUpdate.humidityPoll();
+                connectionManager.sendCommand(humidityUpdate, this);
+
```
done

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,613 @@
+            properties.put("maxTargetTemperature", "30°C / 86°F default");
+        }
+
+        updateProperties(properties);
```
It has helped me answer questions why can't my device do X? Generally the use is early. I learned a few things about my device that wasn't in the manual.  With Zwave, where I'm more familiar, there are all sorts of information that is similar.

apella12 committed:
Address some of the October review comments

Renames energy-related channels (kilowatt-hours, amperes, watts) to energy-consumption, current-draw, and power-consumption across code, documentation, and UI resources for clarity. Refactors hex string conversion to use HexUtils from openHAB core, removing redundant methods from Utils. Improves error handling and introduces LoginFailedException for cloud login failures. Updates related tests and documentation to match new channel names. End of day WIP

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/cloud/Cloud.java:
```diff
@@ -203,7 +206,7 @@ public Cloud(String email, String password, CloudProvider cloudProvider, HttpCli
             if (code != 0) {
                 errMsg = msg;
                 logger.warn("Error {} logging to Cloud: {}", code, msg);
```
Now this logger can be changed to debug as the exception is thrown and handled upstream.

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/discovery/MideaACDiscoveryService.java:
```diff
@@ -0,0 +1,391 @@
+            startDiscoverSocket();
+            // Runs until the socket call gets a time out and throws an exception. When a time out is triggered it means
+            // no data was present and nothing new to discover.
+            while (true) {
```
Good question. 
It triggered me as this while loop is never stopped if there is no device discovered. I watched the super implementation and that also puzzles me.
I'll need to re-visit this.

apella12 replied:
Ok. I have had no problems with the UI scan stopping either with or without finding a(nother) device. I could dig through the logs but also recall a message indicating that discovery stopped, if that would help.

lsiepel reviewed (changes requested):
A follow up to the previous fixes.

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/MideaACConfiguration.java:
```diff
@@ -0,0 +1,146 @@
+    /**
+     * Key and Token Update Frequency in days
+     */
+    public int keyTokenUpdate = 0;
```
I just think someone in the OH community would like to change the token key even if it is unneeded. I have it clearly marked both as days and what zero means
```
Key token update frequency in days
0
Update the Key and Token from the cloud, default 0 to disable. (d)
```

lsiepel replied:
Ok, zero is fine, still hours is more future proof, please adapt.

apella12 replied:
Ok fine, moved to advanced too.

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/Utils.java:
```diff
@@ -0,0 +1,253 @@
+     * @return byte [] array
+     */
+    public static byte[] hexStringToByteArray(String string) {
+        return HexUtils.hexToBytes(string);
```
I think this is resolved then?

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,394 @@
+				<context>password</context>
+				<label>Password</label>
+				<description>Password for cloud account chosen in Cloud Provider.</description>
+				<default>password1</default>
```
As noted elsewhere. This is a live password for the default email address

apella12 committed:
Review update 2. i18n still WIP

Added 'Advanced' column to README and marked advanced parameters in thing-types.xml. Changed keyTokenUpdate unit from days to hours and updated related labels, descriptions, and scheduling logic. Improved MideaACHandler to run discovery and cloud token/key retrieval asynchronously, updating thing status to UNKNOWN during pending operations. Updated i18n strings for clarity and consistency. Changed cloud login error logging from warn to debug.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

apella12 committed:
Improve configuration validation, error handling andi18n messages

Refined configuration parameter validation and error messaging for MideaAC binding. Updated default AC version to 3, clarified keyTokenUpdate interval, improved asynchronous device/cloud parameter retrieval, and enhanced connection error handling. Updated documentation and i18n resources for clearer user guidance.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,389 @@
+			<parameter name="timeout" type="decimal" required="true" min="2" max="10" unit="s">
+				<label>Socket Timeout</label>
+				<description>Connecting socket timeout (seconds). Minimum is 2, maximum is 10 (4 is default).</description>
+				<default>4</default>
```
Should also be marked advanced.

lsiepel commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,634 @@
+    @Override
+    public void initialize() {
+        config = getConfigAs(MideaACConfiguration.class);
+        updateStatus(ThingStatus.UNKNOWN);
```
First check if the configuration is valid, before setting the state to unknown.

apella12 replied:
Would the state be initializing (I found that was not possible to assign). I thought it needed to be something, but will remove and see.

apella12 replied:
What if all config are good, then the thing is never set to anything but online? I thought it needed to be something while going through the initialization()

lsiepel reviewed (changes requested):
I think almost all comments are addressed, my main concern is the initialize method. 
The openHAB model seems stable, that is one of the most important parts.

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,634 @@
+     * 
+     */
+    @Override
+    public void initialize() {
```
TBH last time you helped I almost decided to give up.
All these things need to be done and with the async threads it is happening fast.
I'd like to hold on this for now.

lsiepel replied:
Created the PR here: https://github.com/apella12/openhab-addons/pull/2

Please test and see if all is still working as expected, if so, merge it and we can finish this PR.

apella12 commented on bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java:
```diff
@@ -0,0 +1,634 @@
+                    } catch (Exception e) {
+                        // Required parameter discovery failed
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                                "Could not retrieve required parameters from device or the cloud.");
```
I need an example because these shows as an error in vsc
```
updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.offline.communication_error_token,
                                "Could not retrieve required parameters from device or the cloud.");
```
 and 
```
updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.communication_error_token,
                                "Could not retrieve required parameters from device or the cloud.");
```
with in the i18n
`offline.communication_error_token = Could not retrieve required parameters from device or the cloud.`

message: communication_error_token cannot be resolved or is not a fieldJava(33554502)

apella12 committed:
Improve AC thing initialization documentation and i18n keys

Refactored MideaACHandler initialization to clarify discovery and token/key retrieval steps, using more specific status messages and i18n keys. Updated i18n properties for granular error reporting. Marked 'timeout' and 'version' parameters as advanced in thing-types.xml and set version default to 3. Updated README to reflect 'timeout' as advanced.

Signed-off-by: Bob Eckhoff <katmandodo@yahoo.com>

lsiepel committed:
Refactor initialize

Signed-off-by: Leo Siepel <leosiepel@gmail.com>

lsiepel reviewed (approved):
Thanks, LGTM

apella12 commented:
I have the latest version running. No more testing really needed IMO. Also, the underlying code looked the same, materially, just dispersed around the MideaACHandler class so the initialize() just had method calls.  Thanks for approving !

lsiepel commented:
> o is here.
> I’d like to put this out for new

The thingstatus handling did change a bit but the logic itself was not changed as that was solid. Anyway.

Now, you could add the binding's logo to the openHAB website. See https://www.openhab.org/docs/developer/addons/#add-your-add-on-s-logo-to-the-openhab-website-and-the-ui
